OpenRA.Mods.Common/MapGenerator/Terraformer.cs

```diff
@@ -0,0 +1,1750 @@
+#region Copyright & License Information
+/*
+ * Copyright (c) The OpenRA Developers and Contributors
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using OpenRA.Mods.Common.Terrain;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Primitives;
+using OpenRA.Support;
+using static OpenRA.Mods.Common.Traits.ResourceLayerInfo;
+
+namespace OpenRA.Mods.Common.MapGenerator
+{
+	/// <summary>Collection of high-level map generation utilities.</summary>
+	public class Terraformer
+	{
+		/// <summary>Common denominator for fractional arguments.</summary>
+		public const int FractionMax = 1000;
+
+		/// <summary>Biases or excludes resources at a location during resource planning.</summary>
+		public sealed class ResourceBias
+		{
+			/// <summary>The location of the bias.</summary>
+			public WPos WPos;
+
+			/// <summary>Resources will not be placed within this distance of the actor.</summary>
+			public WDist? ExclusionRadius = null;
+
+			/// <summary>Resources will be biased within this radius.</summary>
+			public WDist? BiasRadius = null;
+
+			/// <summary>
+			/// Biasing function, applied either to all resources or the specific ResourceType.
+			/// Maps the original value and the squared-WDist-from-CPos to a new value.
+			/// </summary>
+			public Func<int, long, int> Bias = null;
+
+			/// <summary>If non-null, encourages resources to become this type.</summary>
+			public ResourceTypeInfo ResourceType = null;
+
+			/// <summary>Create a bias at a location.</summary>
+			public ResourceBias(WPos wpos)
+			{
+				WPos = wpos;
+			}
+
+			/// <summary>Create a bias at an actor's location.</summary>
+			public ResourceBias(ActorPlan actorPlan)
+				: this(actorPlan.WPosCenterLocation)
+			{ }
+		}
+
+		/// <summary>
+		/// Metadata for the values in a CellLayer matching an ID.
+		/// </summary>
+		public sealed class Region
+		{
+			public const int NullId = -1;
+
+			/// <summary>Region ID.</summary>
+			public int Id;
+
+			/// <summary>Area of the region.</summary>
+			public int Area;
+		}
+
+		public enum Side : sbyte
+		{
+			Out = -1,
+			None = 0,
+			In = 1,
+		}
+
+		public static (T[] Types, U[] Weights) SplitDictionary<T, U>(IReadOnlyDictionary<T, U> typeWeights)
```

Comment:
```suggestion
		public static (T[] Types, U[] Weights) SplitAndOrderDictionary<T, U>(IReadOnlyDictionary<T, U> typeWeights)
```

This doesn't look like it belongs in Terraformer

Reply:
It ended up being called from both ExperimentalMapGenerator and Terraformer, so I moved it to Terraformer. Happy to have a more generic location suggested. (Exts?)